sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,912 @@
+//! @brief Solana Rust-based BigNumber for bpf-programs
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use std::fmt;
+
+#[repr(C)]
+#[derive(
+    BorshSerialize,
+    BorshDeserialize,
+    BorshSchema,
+    Clone,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Hash,
+    AbiExample,
+)]
+
+/// BigNumber supports integers greater than
+/// 64 bits for BPF programs
+pub struct BigNumber {
+    value: Vec<u8>,
+    is_negative: bool,
+}
+
+impl BigNumber {
+    /// Returns a BigNumber with initial value of 0
+    pub fn new() -> Self {
+        Self {
+            is_negative: false,
+            value: vec![0u8],
+        }
+    }
+
+    /// Returns the size, in bytes, of BigNum. Typically used in
+    /// assessing buffer size needed to fetch the bytes for serialization, etc.
+    pub fn size_in_bytes(&self) -> usize {
+        self.value.len()
+    }
+
+    /// Returns a BigNumber with initial value set to a u32 value
+    pub fn from_u32(val: u32) -> Result<Self, u64> {
+        if val == 0 {
+            Ok(Self::new())
+        } else {
+            #[cfg(not(target_arch = "bpf"))]
+            {
+                match openssl::bn::BigNum::from_u32(val) {
+                    Ok(bn) => Ok(Self {
+                        is_negative: false,
+                        value: bn.to_vec(),
+                    }),
+                    Err(_) => Err(0x1),
+                }
+            }
+            #[cfg(target_arch = "bpf")]
+            {
+                extern "C" {
+                    fn sol_bignum_from_u32(
+                        bytes_ptr: *mut u8,
+                        bytes_len: *mut u64,
+                        val: u64,
+                    ) -> u64;
```

Comment:
Maybe I  missed it but it doesn't seem that any of these syscalls actually return a meaningful value and instead the syscalls seem to return `Err` which will halt the program.  If you expect all syscall errors to halt there is no reason to pay attention to the result.  Or, you could return u64 values that you can then translate to an error `Enum` in the program space that reflects what the error was.  If you don't expect programs to recover from an error or don't want them to, then don't return an error, just let the programs halt.

Reply:
I will re-assess as I rework the program side error management

Reply:
See above final error handling, basically all errors halt the program as there are no recoverable errors in the maths